File: 104-maximum-depth-of-binary-tree/main.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxDepth(root *TreeNode) int {

    if root == nil {
        return 0
    }

    var dfs func(root *TreeNode) int
    dfs = func(root *TreeNode) int {
        if root == nil {
            return 0
        }

        leftNode := dfs(root.Left)
        rightNode := dfs(root.Right)

        return max(leftNode, rightNode) + 1
    }

    return dfs(root)
    
}

func max(a, b int) int {
    if a < b {
        return b
    }
    return a
}

File: 104-maximum-depth-of-binary-tree/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        
        if not root:
            return 0

        def dfs(node: Optional[TreeNode]) -> int:
            
            if not node:
                return 0

            left_node = dfs(node.left)
            right_node = dfs(node.right)

            return max(left_node, right_node) + 1
        
        return dfs(root)

        
